### PR TITLE
Issues/104

### LINKED ISSUE #104: Log request headers and body at 4xx level when response status is >= 400
Hello, and thanks for the good work! We use and love your package but we do have one problem with it.

We would definitely like to have headers and body of failing requests (status >= 400) to be logged at the level set in `settings.REQUEST_LOGGING_HTTP_4XX_LOG_LEVEL`.

# Current behavior

```py
class LoggingMiddleware(object):
    def __call__(self, request):
        self.process_request(request)
        response = self.get_response(request)
        self.process_response(request, response)
        return response
```

The middleware:

1. Logs incoming request.
2. Lets the response be processed by Django.
3. Logs the response.

This design makes it impossible to log request headers/body at different levels depending on response status code, because, well, response does not exist yet at the time.

# Proposed behavior

Logging both request AND response after response has been processed, allowing customization of log level on both the request and its response.

We would happily offer a pull request for this change :-)

### PR DESCRIPTION
**Pull request by KSD-France**
Fixes #104.

### CODE REVIEW

**Comment by KSD-France** [comment]
@tclancy 

I see 2 possible refactorings of `LoggingMiddleware._log_request_body`:

1. Simplest workaround would be to cache `request.body` before the request is processed and accessing this cached version in `_log_request_body`, yielding same log results as before.
2. More involved refactoring would be to test request's method (`GET`/`POST`/...) and log the cognate `request`'s attribute, *ie*: if processing a `POST` request, log `request.POST` (instead of raw `request.body`).

Your choice!

File: request_logging/middleware.py [hunk 1]
```diff
@@ -239,28 +251,27 @@ def _log_multipart(self, body, logging_context):
...
+                    part = match.expand(r"\1\2/\3\r\n\r\n(binary data)\r\n")
 
             if i != last:
                 part = part + self.boundary
 
             self.logger.log(self.log_level, part, logging_context)
 
     def _log_resp(self, level, response, logging_context):
-        if re.match('^application/json', response.get('Content-Type', ''), re.I):
+        if re.match("^application/json", response.get("Content-Type", ""), re.I):
```
[the remark this answers is not in the document]
Comment (KSD-France) [reply request_logging/middleware.py:265]: Considering that `application/json; charset=UTF-8` is a valid content type (albeit obsolete), the test should probably be: `if response.get("Content-Type", "").lower().startswith("application/json")`.

**Comment by tclancy** [comment]
I think the first is simplest, let's try that.

**Comment by KSD-France** [comment]
OK, done. I think this might be ready for review/merge. Testing on our current project right now.

**Comment by tclancy** [comment]
Let me know when you're done with your testing; things are pretty crazy here this week so I won't have much of a chance to look at it.

**Comment by KSD-France** [comment]
OK, things look pretty smooth so far, let's just give it some time and use and I'll let you know by the end of the week :-)

**Comment by KSD-France** [comment]
Manual testing revealed no issue during the week. I didn't have the opportunity to test with DRF though.

**Comment by KSD-France** [comment]
Nevermind. Just tested on another project that uses DRF and it's all good (both 200 and >= 400 requests).

**Comment by KSD-France** [comment]
Thanks for merging :+1:

**Comment by KSD-France** [comment]
Sorry to revive this PR, just want to know if you have plans to package and deploy to Pypi an updated version including these changes.

**Pull request merged**